src/slurm_plugin/common.py

```diff
@@ -8,5 +8,184 @@
 # or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
+import collections
+import logging
+import re
+import subprocess
+
+import boto3
+from botocore.exceptions import ClientError
+
+from common.schedulers.slurm_commands import update_nodes
+from common.utils import grouper
 
 CONFIG_FILE_DIR = "/opt/parallelcluster/configs/slurm"
+
+log = logging.getLogger(__name__)
+
+
+class InstanceManager:
+    """
+    InstanceManager class.
+
+    Class implementing instance management actions.
+    Should generally be used when launching or terminating instances for slurm integration.
+    """
+
+    class InvalidNodenameError(ValueError):
+        r"""
+        Exception raised when encountering a NodeName that is invalid/incorrectly formatted.
+
+        Valid NodeName format: {queue_name}-{static/dynamic}-{instance_type}-{number}
+        And match: ^([a-z0-9\-_]+)-(static|dynamic)-([a-z0-9-]+.[a-z0-9-]+)-\d+$
+        Sample NodeName: queue1-static-c5.xlarge-2
+        """
+
+        pass
+
+    def __init__(self, region, cluster_name, boto3_config):
+        """Initialize InstanceLauncher with required attributes."""
+        self._region = region
+        self._cluster_name = cluster_name
+        self._boto3_config = boto3_config
+        self.failed_nodes = []
+
+    def clear_failed_nodes(self):
+        """Clear and reset failed nodes list."""
+        self.failed_nodes = []
```

Comment:
why not just doing this as first thing in `add_instances_for_nodes`, without exposing any additional function the caller needs to be aware of?

Reply:
Done